### PR TITLE
[ut] add ut for cache.go, mcs.go, and improve DeleteEndpointSlice

### PR DESCRIPTION
**Pull request by ikaven1024**
Signed-off-by: yingjinhui <yingjinhui@didiglobal.com>

**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**:
Add ut for  pkg/util/helper/cache.go, mcs.go;
And improve DeleteEndpointSlice

**Which issue(s) this PR fixes**:
Part of #2579

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
```release-note
NONE
```

### CODE REVIEW

File: pkg/util/helper/cache_test.go [hunk 1]
```diff
@@ -0,0 +1,304 @@
...
+		},
+		{
+			name: "cluster manager not found",
+			args: args{
+				restMapper: func() meta.RESTMapper {
+					m := meta.NewDefaultRESTMapper([]schema.GroupVersion{corev1.SchemeGroupVersion})
+					m.Add(schema.GroupVersionKind{Version: "v1", Kind: "Pod"}, meta.RESTScopeNamespace)
+					return m
+				}(),
+				manager: func(stopCh <-chan struct{}) genericmanager.MultiClusterInformerManager {
```
Comment (jwcesign) [inline pkg/util/helper/cache_test.go:53]: the way to create `restMapper`  `manager` is quite similar, can we extract it as a common function?

File: pkg/util/helper/mcs_test.go [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+					Port: pointer.Int32(80),
+				}},
+			},
+			wantErr: false,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			if err := CreateOrUpdateEndpointSlice(tt.args.client, tt.args.endpointSlice); (err != nil) != tt.wantErr {
+				t.Errorf("CreateOrUpdateEndpointSlice() error = %v, wantErr %v", err, tt.wantErr)
```
Comment (jwcesign) [inline pkg/util/helper/mcs_test.go:152]: if  want to stop directly, use `t.Fatalf`

File: pkg/util/helper/mcs_test.go [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+				Ports: []discoveryv1.EndpointPort{{
+					Port: pointer.Int32(80),
+				}},
+			},
+			wantErr: false,
+		},
+		{
+			name: "exist and not update",
+			args: args{
+				client: fake.NewClientBuilder().WithScheme(gclient.NewSchema()).WithObjects(
```
Comment (jwcesign) [inline pkg/util/helper/mcs_test.go:103]: same with before, can we extract functions？

File: pkg/util/helper/mcs_test.go (hunk 2 above)
Comment (ikaven1024) [reply pkg/util/helper/mcs_test.go:152]: Golang generate the test  function with `t.Error`. I don't modify it.

Comment (jwcesign) [reply pkg/util/helper/mcs_test.go:152]: I check the usage of Errorf and Fatalf: https://ieftimov.com/posts/testing-in-go-failing-tests/#when-to-use-errorf-or-fatalf

File: pkg/util/helper/mcs_test.go (hunk 3 above)
Comment (ikaven1024) [reply pkg/util/helper/mcs_test.go:103]: Test objects have tiny differences.  Define it in the case, easy to read.

File: pkg/util/helper/cache_test.go [hunk 4]
```diff
@@ -0,0 +1,304 @@
...
+			args: args{
+				restMapper: func() meta.RESTMapper {
+					m := meta.NewDefaultRESTMapper([]schema.GroupVersion{corev1.SchemeGroupVersion})
+					m.Add(corev1.SchemeGroupVersion.WithKind("Pod"), meta.RESTScopeNamespace)
+					return m
+				}(),
+				manager: func(stopCh <-chan struct{}) genericmanager.MultiClusterInformerManager {
+					m := genericmanager.NewMultiClusterInformerManager(stopCh)
+					m.ForCluster("cluster", fake.NewSimpleDynamicClient(scheme.Scheme), 0).
+						Lister(corev1.SchemeGroupVersion.WithResource("pods"))
```
Comment (XiShanYongYe-Chang) [inline pkg/util/helper/cache_test.go:121]: Just a question: why need to call `Lister`, I try with removing it and it works ok.

File: pkg/util/helper/cache_test.go [hunk 5]
```diff
@@ -0,0 +1,304 @@
...
+				restMapper: func() meta.RESTMapper {
+					m := meta.NewDefaultRESTMapper([]schema.GroupVersion{corev1.SchemeGroupVersion})
+					m.Add(corev1.SchemeGroupVersion.WithKind("Pod"), meta.RESTScopeNamespace)
+					return m
+				}(),
+				manager: func(stopCh <-chan struct{}) genericmanager.MultiClusterInformerManager {
+					m := genericmanager.NewMultiClusterInformerManager(stopCh)
+					m.ForCluster("cluster", fake.NewSimpleDynamicClient(scheme.Scheme,
+						&corev1.Pod{ObjectMeta: metav1.ObjectMeta{Name: "pod", Namespace: "default"}},
+					), 0).Lister(corev1.SchemeGroupVersion.WithResource("pods"))
```
Comment (XiShanYongYe-Chang) [inline pkg/util/helper/cache_test.go:145]: Same as the question above.

**Review by XiShanYongYe-Chang — COMMENTED** [review]
Other parts LGTM, thanks~

File: pkg/util/helper/cache_test.go (hunk 5 above)
Comment (ikaven1024) [reply pkg/util/helper/cache_test.go:145]: `Lister` is called to register an informer. Without it, informer will not create, and fetch resource from server, not from cache.
https://github.com/karmada-io/karmada/blob/5811aea95303236ff1c98eb110cea065b4020143/pkg/util/helper/cache.go#L42-L44

Comment (XiShanYongYe-Chang) [reply pkg/util/helper/cache_test.go]: Thanks for explaining!

**Comment by jwcesign** [comment]
/lgtm
cc @RainbowMango PTAL

**Review by RainbowMango — APPROVED** [review]
/approve

**Pull request merged**